File: apps/android/feature/presence/src/main/kotlin/space/fishhub/android/feature/presence/PresenceFormatter.kt
package space.fishhub.android.feature.presence

import android.content.Context
import android.text.format.DateFormat
import space.fishhub.android.data.presence.PresenceDisplayStatus
import space.fishhub.android.data.presence.PresencePreference
import space.fishhub.android.data.presence.PresenceSnapshot
import space.fishhub.android.data.presence.PresenceStatus
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import java.util.Locale
import kotlin.math.max

class PresenceFormatter(
    context: Context? = null,
    private val locale: () -> Locale = { Locale.getDefault() },
    private val zoneId: () -> ZoneId = { ZoneId.systemDefault() },
    private val is24Hour: () -> Boolean = {
        context?.let { DateFormat.is24HourFormat(it.applicationContext) } ?: false
    },
) {
    fun format(
        snapshot: PresenceSnapshot?,
        nowMs: Long,
        ownPreference: PresencePreference? = null,
        ownPreferenceExpired: Boolean = false,
    ): PresencePresentation {
        if (ownPreference != null && !ownPreferenceExpired) {
            when (ownPreference) {
                PresencePreference.Away -> return fixed(PresenceDisplayStatus.Away)
                PresencePreference.Busy -> return fixed(PresenceDisplayStatus.Busy)
                PresencePreference.Invisible -> return fixed(PresenceDisplayStatus.Invisible)
                PresencePreference.Automatic -> Unit
            }
        }
        val heartbeat = snapshot?.lastHeartbeatAt?.toInstantOrNull()?.toEpochMilli()
        val stale = snapshot?.status != null &&
            snapshot.status != PresenceStatus.Offline &&
            (heartbeat == null || heartbeat < nowMs - StaleAfterMs)
        val status = if (stale) PresenceStatus.Offline else snapshot?.status ?: PresenceStatus.Offline
        val display = when (status) {
            PresenceStatus.Online -> PresenceDisplayStatus.Online
            PresenceStatus.Idle -> PresenceDisplayStatus.Idle
            PresenceStatus.Away -> PresenceDisplayStatus.Away
            PresenceStatus.Busy -> PresenceDisplayStatus.Busy
            PresenceStatus.Offline -> PresenceDisplayStatus.Offline
        }
        val detail = if (display == PresenceDisplayStatus.Offline) {
            snapshot?.lastSeenAt?.let { formatLastSeen(it, nowMs) }
        } else {
            null
        }
        return PresencePresentation(display, label(display), detail)
    }

    fun label(status: PresenceDisplayStatus): String = when (status) {
        PresenceDisplayStatus.Online -> "Online"
        PresenceDisplayStatus.Idle -> "Idle"
        PresenceDisplayStatus.Away -> "Away"
        PresenceDisplayStatus.Busy -> "Do not disturb"
        PresenceDisplayStatus.Invisible -> "Invisible"
        PresenceDisplayStatus.Offline -> "Offline"
    }

    private fun fixed(status: PresenceDisplayStatus) =
        PresencePresentation(status = status, label = label(status))

    private fun formatLastSeen(value: String, nowMs: Long): String? {
        val instant = value.toInstantOrNull() ?: return null
        val seenMs = instant.toEpochMilli()
        val elapsed = max(0, nowMs - seenMs)
        if (elapsed < HourMs) {
            val minutes = max(1, elapsed / MinuteMs)
            return "Last seen $minutes ${if (minutes == 1L) "minute" else "minutes"} ago"
        }
        val zone = zoneId()
        val nowDate = Instant.ofEpochMilli(nowMs).atZone(zone).toLocalDate()
        val seenDateTime = instant.atZone(zone)
        if (elapsed < DayMs && seenDateTime.toLocalDate() == nowDate) {
            val hours = max(1, elapsed / HourMs)
            return "Last seen $hours ${if (hours == 1L) "hour" else "hours"} ago"
        }
        if (seenDateTime.toLocalDate() == nowDate.minusDays(1)) {
            val pattern = if (is24Hour()) "HH:mm" else "h:mm a"
            val time = DateTimeFormatter.ofPattern(pattern, locale()).format(seenDateTime)
            return "Last seen yesterday at $time"
        }
        val date = DateTimeFormatter.ofLocalizedDate(FormatStyle.MEDIUM)
            .withLocale(locale())
            .format(seenDateTime)
        return "Last seen on $date"
    }

    private fun String.toInstantOrNull(): Instant? = runCatching { Instant.parse(this) }.getOrNull()

    private companion object {
        const val MinuteMs = 60_000L
        const val HourMs = 60 * MinuteMs
        const val DayMs = 24 * HourMs
        const val StaleAfterMs = 90_000L
    }
}

File: apps/ios/FishKit/Sources/Presence/Logic/PresenceFormatter.swift
import Foundation
import PresenceData
import UIComponents

/// Pure presentation rules — the port of the web `getPresencePresentation`
/// and Android's `PresenceFormatter`. The owner's manual preference wins
/// while unexpired; a non-offline snapshot whose heartbeat is missing or
/// older than 90 seconds renders offline (the backend coalesces heartbeats
/// for up to 60 seconds, so the threshold must stay at 90); last-seen copy
/// appears only for offline subjects with an unsanitized timestamp.
public struct PresenceFormatter: Sendable {
    private let locale: Locale
    private let calendar: Calendar
    private let timeZone: TimeZone

    public init(
        locale: Locale = .current,
        calendar: Calendar = .current,
        timeZone: TimeZone = .current
    ) {
        self.locale = locale
        var localized = calendar
        localized.timeZone = timeZone
        localized.locale = locale
        self.calendar = localized
        self.timeZone = timeZone
    }

    // MARK: - Labels

    public static func label(for status: PresenceDisplayStatus) -> String {
        switch status {
        case .online: "Online"
        case .idle: "Idle"
        case .away: "Away"
        case .busy: "Do not disturb"
        case .invisible: "Invisible"
        case .offline: "Offline"
        }
    }

    // MARK: - Subjects

    public func format(_ snapshot: PresenceSnapshot?, now: Date) -> PresencePresentation {
        let status = effectiveStatus(of: snapshot, now: now)
        let detail = status == .offline
            ? snapshot?.lastSeenAt.flatMap { lastSeenCopy($0, now: now) }
            : nil
        return PresencePresentation(
            status: status,
            label: Self.label(for: status),
            detail: detail
        )
    }

    // MARK: - Own status

    /// The owner's display derives preference-first: Away, Do not disturb,
    /// and Invisible show immediately even while the snapshot lags;
    /// Automatic falls through to the snapshot-derived status.
    public func formatOwn(
        snapshot: PresenceSnapshot?,
        preference: PresencePreferenceSetting,
        now: Date
    ) -> PresencePresentation {
        let effective = expired(preference, now: now)
            ? PresencePreferenceSetting()
            : preference
        switch effective.preference {
        case .away:
            return PresencePresentation(status: .away, label: Self.label(for: .away))
        case .busy:
            return PresencePresentation(status: .busy, label: Self.label(for: .busy))
        case .invisible:
            return PresencePresentation(
                status: .invisible,
                label: Self.label(for: .invisible)
            )
        case .automatic:
            return format(snapshot, now: now)
        }
    }

    public func expired(_ preference: PresencePreferenceSetting, now: Date) -> Bool {
        guard let expiresAt = preference.expiresAt,
              let expiry = PresenceTimestamp.parse(expiresAt) else { return false }
        return expiry <= now
    }

    // MARK: - Rules

    private func effectiveStatus(
        of snapshot: PresenceSnapshot?,
        now: Date
    ) -> PresenceDisplayStatus {
        guard let snapshot else { return .offline }
        guard snapshot.status != .offline else { return .offline }
        guard let heartbeat = PresenceTimestamp.parse(snapshot.lastHeartbeatAt),
              heartbeat >= now.addingTimeInterval(-Self.staleSeconds) else {
            return .offline
        }
        switch snapshot.status {
        case .online: return .online
        case .idle: return .idle
        case .away: return .away
        case .busy: return .busy
        case .offline: return .offline
        }
    }

    private func lastSeenCopy(_ value: String, now: Date) -> String? {
        guard let seen = PresenceTimestamp.parse(value) else { return nil }
        let elapsed = max(0, now.timeIntervalSince(seen))
        if elapsed < Self.hourSeconds {
            let minutes = max(1, Int(elapsed / 60))
            return "Last seen \(minutes) \(minutes == 1 ? "minute" : "minutes") ago"
        }
        let sameLocalDay = calendar.isDate(seen, inSameDayAs: now)
        if elapsed < Self.daySeconds, sameLocalDay {
            let hours = max(1, Int(elapsed / Self.hourSeconds))
            return "Last seen \(hours) \(hours == 1 ? "hour" : "hours") ago"
        }
        if isYesterday(seen, relativeTo: now) {
            return "Last seen yesterday at \(timeOfDay(seen))"
        }
        return "Last seen on \(calendarDate(seen))"
    }

    private func isYesterday(_ seen: Date, relativeTo now: Date) -> Bool {
        let days = calendar.dateComponents(
            [.day],
            from: calendar.startOfDay(for: seen),
            to: calendar.startOfDay(for: now)
        ).day
        return days == 1
    }

    private func timeOfDay(_ date: Date) -> String {
        date.formatted(Date.FormatStyle(
            date: .omitted,
            time: .shortened,
            locale: locale,
            calendar: calendar,
            timeZone: timeZone
        ))
    }

    private func calendarDate(_ date: Date) -> String {
        date.formatted(Date.FormatStyle(
            date: .abbreviated,
            time: .omitted,
            locale: locale,
            calendar: calendar,
            timeZone: timeZone
        ))
    }

    private static let staleSeconds =
        TimeInterval(PresenceRules.staleAfter.components.seconds)
    private static let hourSeconds: TimeInterval = 3_600
    private static let daySeconds: TimeInterval = 86_400
}
